File: benchmarks/tasks/compose_chain/compose_chain.hs
-- SPDX-License-Identifier: MIT-0
module Main where

double :: Int -> Int
double x = x * 2

inc :: Int -> Int
inc x = x + 1

square :: Int -> Int
square x = x * x

transform :: Int -> Int
transform = square . inc . double

showList' :: Show a => [a] -> String
showList' xs = "[" ++ unwords (map show xs) ++ "]"

main :: IO ()
main = putStrLn (showList' (map transform [1..5]))

File: benchmarks/tasks/compose_chain/compose_chain.py
# SPDX-License-Identifier: MIT-0

def double(x): return x * 2
def inc(x): return x + 1
def square(x): return x * x

def compose(*fns):
    def composed(x):
        for f in fns:
            x = f(x)
        return x
    return composed

pipeline = compose(double, inc, square)
result = [pipeline(x) for x in [1, 2, 3, 4, 5]]
print("[" + " ".join(str(x) for x in result) + "]")
